Mathlib/Data/Nondet/Basic.lean

```diff
@@ -0,0 +1,170 @@
+  bind := bind
+
+/-- `Nondet m` is an alternative monad. -/
+instance : Alternative (Nondet m) where
+  failure := .nil
+  orElse x y := .mk <| x.data.append fun _ => (y ()).data
+
+instance : MonadLift m (Nondet m) where
+  monadLift := singletonM
+
+/--
+Lift a list of monadic values to a nondeterministic value.
+We ensure that each monadic value is evaluated with the same backtrackable state.
+-/
+def ofListM (L : List (m α)) : Nondet m α :=
+  .squash fun _ => do
+    let s ← saveState
+    return .mk <| MLList.ofMLList <| L.map fun x => do
+      restoreState s
+      let a ← x
+      pure (← saveState, a)
+
+/--
+Lift a list of values to a nondeterministic value.
+(The backtrackable state in each will be identical:
+whatever the state was when we first read from the result.)
+-/
+def ofList [Alternative m] (L : List α) : Nondet m α := ofListM (L.map pure)
+
+/-- Apply a function which returns values in the monad to every alternative of a `Nondet m α`. -/
+partial def mapM (f : α → m β) (L : Nondet m α) : Nondet m β :=
+  L.bind fun a => singletonM (f a)
+
+/-- Apply a function to each alternative in a `Nondet m α` . -/
+def map (f : α → β) (L : Nondet m α) : Nondet m β :=
+  L.mapM fun a => pure (f a)
+
+/-- Convert a monadic optional value to a nondeterministic value. -/
+def ofOptionM (x : m (Option α)) : Nondet m α := .squash fun _ => do
+  match ← x with
+  | none => return .nil
+  | some a => return singleton a
+
+/-- Convert an optional value to a nondeterministic value. -/
+def ofOption (x : Option α) : Nondet m α := ofOptionM (pure x)
+
+/-- Filter and map a nondeterministic value using a monadic function which may return `none`. -/
+partial def filterMapM (f : α → m (Option β)) (L : Nondet m α) : Nondet m β :=
+  L.bind fun a => ofOptionM (f a)
+
+/-- Filter and map a nondeterministic value. -/
+def filterMap (f : α → Option β) (L : Nondet m α) : Nondet m β :=
+  L.filterMapM fun a => pure <| (f a)
+
+/-- Filter a nondeterministic value using a monadic predicate. -/
+partial def filterM (p : α → m (ULift Bool)) (L : Nondet m α) : Nondet m α :=
+  L.filterMapM fun a => do
+    if (← p a).down then
+      pure (some a)
+    else
+      pure none
+
+/-- Filter a nondeterministic value. -/
+def filter (p : α → Bool) (L : Nondet m α) : Nondet m α :=
+  L.filterM fun a => pure <| .up (p a)
+
+/--
+Find the first alternative in a nondeterministic value, as a monadic value.
+-/
+def head [Alternative m] (L : Nondet m α) : m α := (·.2) <$> MLList.head L.data
+
+/--
+Find the value of a monadic function on the first alternative in a nondeterministic value
+where the function succeeds.
+-/
+def firstM [Alternative m] (L : Nondet m α) (f : α → m (Option β)) : m β :=
+  L.filterMapM f |>.head
+
+/--
+Convert a non-deterministic value into a lazy list, keeping the backtrackable state.
+Be careful that monadic operations on the `MLList` will not respect this state!
+-/
+def toMLList' (L : Nondet m α) : MLList m (σ × α) := L.data
```

Comment:
Is there a reason not to merge this with `.data`? Having two names for the same thing doesn't seem very helpful.

If you do keep this,

```suggestion
abbrev toMLList' (L : Nondet m α) : MLList m (σ × α) := L.data
```

Reply:
I would suggest renaming `data` to `toMLList` and then renaming `toMLList` to `toMLList'`. I'd usually expect the primed version to do some more discarding, like with `run` in reader/state monads.

Reply:
I think a name that indicates what the MLList _represents_ would be better than eitehr `data` or `toMLList`

Reply:
Do you have a suggestion for such a name? (Do you at least agree that toMLList is less generic than data?)

Reply:
I don't have a good enough understanding of this new monad to know what a good name would be I'm afraid; nor do I have a preference, but I'm happy to go with your opinion.